File: code/py/ch14/14.02.py
import pandas as pd
my_data = pd.DataFrame(
    {'x': [0, -16, 10, 10],
     'y': [0, 0, 10, -15]},
    index=['A', 'B', 'C', 'D'])

from scipy.spatial import distance

# ユークリッド距離
distance.cdist(my_data, my_data)
# あるいは
distance.cdist(my_data, my_data, metric='euclidean')
#> array([[ 0.        , 16.        , 14.14213562, 18.02775638],
#>        [16.        ,  0.        , 27.85677655, 30.01666204],
#>        [14.14213562, 27.85677655,  0.        , 25.        ],
#>        [18.02775638, 30.01666204, 25.        ,  0.        ]])

# マンハッタン距離
distance.cdist(my_data, my_data, metric='cityblock')
#> array([[ 0., 16., 20., 25.],
#>        [16.,  0., 36., 41.],
#>        [20., 36.,  0., 25.],
#>        [25., 41., 25.,  0.]])

#### 14.2.2.2 Pythonの場合

import pandas as pd
my_data = pd.DataFrame(
    {'x': [0, -16, 10, 10],
     'y': [0, 0, 10, -15]},
    index=['A', 'B', 'C', 'D'])

# クラスター分析
from scipy.cluster import hierarchy
my_result = hierarchy.linkage(my_data)
# あるいは
my_result = hierarchy.linkage(my_data, metric='euclidean', method='complete')

# デンドログラムの描画
hierarchy.dendrogram(my_result, labels=my_data.index)

# クラスター数を3にする場合に，各点がどのクラスターに属しているかを確認する．
hierarchy.cut_tree(my_result, 3)
#> array([[0],
#>        [1],
#>        [0],
#>        [2]])

#### 14.2.3.2 Pythonの場合

import pandas as pd
my_data = pd.DataFrame(
    {'language': (  0, 20, 20, 25, 22, 17),
     'english':  (  0, 20, 40, 20, 24, 18),
     'math':     (100, 20,  5, 30, 17, 25),
     'science':  (  0, 20,  5, 25, 16, 23),
     'society':  (  0, 20, 30,  0, 21, 17)},
    index=['A', 'B', 'C', 'D', 'E', 'F'])

import seaborn as sns
sns.clustermap(my_data,
               z_score=1, # 列ごとの標準化
               metric='euclidean',
               method='complete')

#### 14.2.4.2 Pythonの場合

import pandas as pd
my_data = pd.DataFrame(
    {'x': [0, -16, 10, 10],
     'y': [0, 0, 10, -15]},
    index=['A', 'B', 'C', 'D'])

# クラスター数を3として，非階層的クラスター分析を行う．
from sklearn.cluster import KMeans
my_result = KMeans(n_clusters=3).fit(my_data)

# 各データがどのクラスターに属しているかを確認する．
my_result.labels_
#> array([1, 0, 1, 2], dtype=int32)

# アヤメのデータ
import statsmodels.api as sm
iris = sm.datasets.get_rdataset('iris', 'datasets').data
my_data = iris.iloc[:, 0:4]

# 階層的クラスター分析
#from scipy.cluster import hierarchy
#my_hc = hierarchy.linkage(my_data)
#my_data.index = hierarchy.cut_tree(my_hc, 3)[:,0] # 結果の保存

# 非階層的クラスター分析
from sklearn.cluster import KMeans
my_data.index = KMeans(n_clusters=3).fit(my_data).labels_ # 結果の保存

# 主成分分析
from pca import pca
my_model = pca(n_components=4)
my_result = my_model.fit_transform(my_data)

# バイプロット
my_model.biplot()

File: code/R/ch14/14.02.R
### 14.2.1 距離

my_data <- data.frame(
  x = c(  0, -16,  10,  10),
  y = c(  0,   0,  10, -15),
  row.names = c("A", "B", "C", "D"))

# ユークリッド距離
my_data %>% dist
# あるいは
my_data %>% dist("euclidian")

#>          A        B        C
#> B 16.00000                  
#> C 14.14214 27.85678         
#> D 18.02776 30.01666 25.00000

# マンハッタン距離
my_data %>% dist("manhattan")
#>    A  B  C
#> B 16      
#> C 20 36   
#> D 25 41 25

#### 14.2.2.1 Rの場合

library(tidyverse)
my_data <- data.frame(
  x = c(  0, -16,  10,  10),
  y = c(  0,   0,  10, -15),
  row.names = c("A", "B", "C", "D"))

# 距離の計算
my_dist <- my_data %>% dist
# あるいは
my_dist <- my_data %>% dist("euclidian")

# 階層的クラスター分析
my_result <- my_dist %>% hclust
# あるいは
my_result <- my_dist %>% hclust("complete")

plot(my_result, hang = -1)    # デンドログラム
rect.hclust(my_result, k = 3) # クラスター（3個）の明示
# あるいは
factoextra::fviz_dend(my_result, # デンドログラム（図14.3の左）
                      k = 3,     # クラスター（3個）の明示
                      rect = T, rect_fill = T) # クラスターを囲む長方形

# クラスター数を3にする場合に，各点がどのクラスターに属しているかを確認する．
my_result %>% cutree(3)
#> A B C D 
#> 1 2 1 3 

#### 14.2.3.1 Rの場合

library(tidyverse)
my_data <- data.frame(
  language = c(  0, 20, 20, 25, 22, 17),
  english  = c(  0, 20, 40, 20, 24, 18),
  math     = c(100, 20,  5, 30, 17, 25),
  science  = c(  0, 20,  5, 25, 16, 23),
  society  = c(  0, 20, 30,  0, 21, 17),
  row.names = c("A", "B", "C", "D", "E", "F"))

my_data %>%
  scale %>%          # 標準化，
  as.matrix %>%      # 行列にしてから，
  gplots::heatmap.2( # 可視化
    distfun = function(m) { dist(x = m, method = 'euclidean') },
    hclustfun = function(d) { hclust(d = d, method = 'complete') },
    cexRow = 1, # 行ラベルのサイズ
    cexCol = 1) # 列ラベルのサイズ

#### 14.2.4.1 Rの場合

my_data <- data.frame(
  x = c(  0, -16,  10,  10),
  y = c(  0,   0,  10, -15),
  row.names = c("A", "B", "C", "D"))

# クラスター数を3として，非階層的クラスター分析を行う．
my_result <- my_data %>% kmeans(3)

# 各データがどのクラスターに属しているかを確認する．
my_cluster <- my_result$cluster
my_cluster
#> A B C D 
#> 2 3 2 1 

### 14.2.5 主成分分析とクラスター分析

library(tidyverse)
# アヤメのデータ
my_data <- iris[, -5]

# 主成分分析
my_result <- prcomp(my_data)

# 階層的クラスター分析
my_cluster <- my_data %>% scale %>% dist %>% hclust %>% cutree(3)

# 非階層的クラスター分析
#my_cluster <- (my_data %>% scale %>% kmeans(3))$cluster

# バイプロット
my_result %>%
  ggbiplot::ggbiplot(
    scale = 0,
    groups = as.factor(my_cluster)) # クラスターによるグループ分け

